chef/automate
# [Automate-3385] Policy dropdown for tokens

### :nut_and_bolt: Description: What code changed, and why?

This PR provides a valuable exercise in applying [SOLID](https://en.wikipedia.org/wiki/SOLID) principles.

As shown, the ProjectsDropdownComponent is used directly in 3 places: CreateObjectModalComponent, TeamDetailsComponent, and ApiTokenDetailsComponent,
while the CreateObjectModal in turn is used in 3 places: TeamManagement, ApiTokenList, and ProjectList. The newly introduced pieces are in blue.

![image](https://user-images.githubusercontent.com/6817500/80418362-25406e80-888c-11ea-942a-adaa6ad2a676.png)

#### Single Responsibility Principle (SRP)

Previously, each of those components did some of the work that should have been encapsulated within the ProjectsDropdownComponent itself, namely:
(a) Generating the list of assignable projects and passing it through to the ProjectsDropdownComponent.  However, that list just came from the ngrx/store, so it is just as accessible to the ProjectsDropdownComponent to do itself, with the added benefit of (1) encapsulating that responsibility where it should be and (2) eliminating multiple occurrences of repeated code.
(b) Enabling and disabling the ProjectsDropdownComponent, based on the list of assignable projects. It is the ProjectsDropdownComponent itself that knows that list, so it can manage that itself.
(c) Reacting to each check or uncheck action within the dropdown. Clearly, that is the responsibility of the ProjectsDropdownComponent, not the calling components; they only need to know when the user has finished within the dropdown, identified by when the dropdown closes.

The above led to eliminating three of the four ProjectsDropdownComponent parameters and adding two new ones (with the net result of one less parameter total).

#### Open-Closed Principle (OCP)

The original ProjectsDropdownComponent could have been modified to handle policies as well as projects by tweaking the code with some conditionals and a few extra functions.  But a better approach was to genericize the component so it does not care if it is handling policies or projects.  This new component (in blue) is ResourceDropdownComponent. ProjectsDropdownComponent, which knows about projects, has now become a subclass of ResourceDropdownComponent. A second subclass now knows about policies. In most languages, subclassing always deals with first-class objects. But Angular is unique in that you could do that (as with ProjectsDropdownComponent), but you can also do "inline" subclasses in a template. It made sense to do that because (a) it is only needed once, inside CreateObjectModal, and (b) it does not require any custom code, just setting parameters.  In the diagram, you will observe the two subclasses indicated by the dotted lines.

#### Liskov substitution principle (LSP)

To make the above sub-classing work, ResourceDropdownComponent needed to be able to handle a common data type. This relied on the fact that even though policies and projects are quite different objects, they share the commonalities needed: `name` (for display) and `id` (for processing), plus the extension required for the dropdowns, the `checked` property.  So the ProjectsDropdownComponent manages a list of projects (extended with `checked`) while the CreateObjectModalComponent manages a list of policies (also extended with the `checked` property).  Thanks to covariance, we can have a parameter `resources` (an array of `ResourceChecked` in ResourceDropdownComponent) that is a receptacle for either `ProjectChecked` or `PolicyChecked` inputs provided by its subclasses. That is, it is OK to have more properties than `ResourceChecked` requires; those just sit on the sidelines while we work with the required properties.

To be done:
1. Permissions around the dropdown.
2. Organizing by chef managed and custom within the dropdown.
3. UX work: need to notify user that policies, unlike projects, are not part of the token, and you cannot go into the details to edit them or even see them.

### :chains: Related Resources
NA

### :+1: Definition of Done
The dialog opens upon selecting <kbd>Create Token</kbd> on Tokens page; teams and projects dialogs (using the same modal) are unchanged.

![image](https://user-images.githubusercontent.com/6817500/80421686-bb2ac800-8891-11ea-97b6-23e6c5d486d0.png)

(1) Opening the modal shows all available policies (currently sorted by name; that will be enhanced in a follow-up).
(2) Entering a string that occurs in some policy names, filters the list to that set of names.
(3) Entering a string that does not occur shows the message indicated.
![image](https://user-images.githubusercontent.com/6817500/80419340-cda30280-888d-11ea-902e-31cdded8d301.png)

Selecting one policy mirrors that policy name in the dropdown label; selecting more than one indicates that number. 
![image](https://user-images.githubusercontent.com/6817500/80419755-8c5f2280-888e-11ea-9413-be9e93038c8c.png)

Upon selecting <kbd>Create Token</kbd> from the modal, notifications occur for creating the token plus for each amended policy.
<img width="1056" alt="image" src="https://user-images.githubusercontent.com/6817500/80420299-64bc8a00-888f-11ea-83ef-c6e112752c0e.png">
(Note that I modified the AddMember notification to include the policy id; even with just a single policy, it enhances the meaning.)

Finally, note that if you go in to edit a token or team and change the contents of the projects dropdown, it will enable the Save button. If you then change it back to its original state, it will disable the Save button.

### :athletic_shoe: How to Build and Test the Change
1. Rebuild automate-ui
2. Go to Settings >> Tokens
3. Add a new token to see the **_policy_** dropdown

Important: to confirm that the **_project_** dropdown still works, visit the tokens and teams list pages to create them, plus the tokens and teams details pages to update them.

### :white_check_mark: Checklist
